Refactor bagit serialization to match new specification

ThomasThelen opened the pull request:
## Background
Pull request #112 enabled users to add file location information about a data object. It also placed that file in the appropriate directory (somewhere in `./data/`) when serializing to bagit. The PR however didn't refactor the rest of the serialization code to match the new format that Metacat uses.

## Survey of Changes

### DataPakage::serializeToBagit

#### Write to Bag Files on The Fly
Previously, tag file information (manifest, tagmanifest, etc) were stored in objects and appended as package files were written. I've changed this to write to the files on the fly, as the data is generated. You'll see a few more references to file writing and you'll also see that I removed the objects like
```
  pidMap <- character()	
  manifest <- character()
```

#### Removal of pid-mapping.txt
This file is no longer supported in the current specification so I removed it

#### Writing System Metadata to The Bag
The current spec wants system metadata in `./metadata/sysmeta` so you'll see some changes that reference system metadata.

#### Renaming Resource Map
The resource map now has the name `oai-ore.xml` and lives in `./metadata`

#### Writing Science Metadata to Bag
The science metadata is now included in the exported bag. You'll note that I _only_ added support for EML. The problem lies with finding a way to identify which DataObject is the science metadata document. The ideal way (and only way I thought of) 
 is to ask the coordinating node for the format list, and then look for all of the ones that have type `METADATA` and check the format type. It seemed a little off to have to make a request to the CN every time a bag is created (and took a noticeable amount of time when I tested it) , so I'm open to caching suggestions.

### DataPackage::write_to_bag
Note that this method shouldn't be exposed to users so I followed the pattern of not exporting it.

This logic was originally lifted from [this bit of code](https://github.com/ropensci/datapack/blob/master/R/DataPackage.R#L1663). Since we now write system metadata, the EML, etc, I decided to pull it out into a new method. If wanted, I can put this in a lambda to keep people from accessing it.

The arguments to this method are probably them most confusing part of this PR. The reason for the two flags....

System metadata are processed differently than ordinary data files so there's a flag to dictate that behaviour. The science metadata is _like_ system metadata in the sense that it belongs in the tagmanifest, but also acts like a data object in the sense that it uses the same method, `getData`, to get the bytes. Duck typing SystemMetadata to follow the DataObject would fix this, but I think the current workaround isn't horribly messy. 

### Unit Tests
The original serialize test was written in the context of provenance (variable names, prov was added, etc). I removed this code from the test since it had no effect on the actual bag. I also renamed some of the variables to reflect this. I beefed up the testing by checking the contents of the tagmanifest & manifest files to make sure they match what's in the payload & tag directories.

gothub commented:
@ThomasThelen you might consider using the RDF relationships in the resource map to determine what the metadata object is for a package. The metadata object will be the subject of the `documents` relationship, or the object of the `isDocumentedBy` relationship, for example, with the metadata document with pid `doi:10.6085/AA/marine_ltm.9.1`:
```
  <rdf:Description rdf:about="https://cn.dataone.org/cn/v1/resolve/doi:10.6085%2FAA%2Fmarine_ltm.9.1">
    <ore:isAggregatedBy rdf:datatype="http://www.w3.org/2001/XMLSchema#anyURI">https://cn.dataone.org/cn/v1/resolve/resourceMap_marine_ltm.9.1#aggregation</ore:isAggregatedBy>
    <dcterms:identifier rdf:datatype="http://www.w3.org/2001/XMLSchema#string">doi:10.6085/AA/marine_ltm.9.1</dcterms:identifier>
    <j.0:documents rdf:resource="https://cn.dataone.org/cn/v1/resolve/doi:10.6085%2FAA%2Fmarine_ltm.10.1"/>
  </rdf:Description>
``` 
and
```
  <rdf:Description rdf:about="https://cn.dataone.org/cn/v1/resolve/doi:10.6085%2FAA%2Fmarine_ltm.10.1">
    <ore:isAggregatedBy rdf:datatype="http://www.w3.org/2001/XMLSchema#anyURI">https://cn.dataone.org/cn/v1/resolve/resourceMap_marine_ltm.9.1#aggregation</ore:isAggregatedBy>
    <dcterms:identifier rdf:datatype="http://www.w3.org/2001/XMLSchema#string">doi:10.6085/AA/marine_ltm.10.1</dcterms:identifier>
    <j.0:isDocumentedBy rdf:resource="https://cn.dataone.org/cn/v1/resolve/doi:10.6085%2FAA%2Fmarine_ltm.9.1"/>
  </rdf:Description>
```

btw - searching for metadata objects with formatId='eml://ecoinformatics' misses EML 2.2.0 objects, as the new formatId for EML 2.2.0 has been updated to 'https://eml.ecoinformatics.org/eml-2.2.0'.

How would you like to proceed with the PR?
This scheme should work for any metadata format id in DataONE.

BTW - searching for

amoeba commented on R/DataPackage.R:
```diff
@@ -1588,161 +1588,134 @@ setMethod("serializeToBagIt", signature("DataPackage"), function(x, mapId=NA_cha
-        relFile <- paste(relFile, getIdentifier(dataObj), sep="")
+    # Create a temp working area where the BagIt directory structure will be created
+    tmpDir <- tempdir()
+    bagDir <- sprintf("%s/bag", tmpDir)
```
Hey, just a note I had while scanning. I think the package was already doing this before this PR but I see a lot of places where filesystem paths are constructed with strings separated by "/" chars. It's better to use `file.path` so our packages remain cross-platform. Would you be able to update the relevant places in your PR to do things that way?

ThomasThelen replied:
Ah good catch. I'll definitely get that change in.

ThomasThelen replied:
Done in 4ca63f2ceb0cb8f73d07700053b23926407cab35

amoeba replied:
Nice, thanks!

ThomasThelen commented:
> @ThomasThelen you might consider using the RDF relationships in the resource map to determine what the metadata object is for a package. The metadata object will be the subject of the `documents` relationship, or the object of the `isDocumentedBy` relationship, for example, with the metadata document with pid `doi:10.6085/AA/marine_ltm.9.1`:
> 
> ```
>   <rdf:Description rdf:about="https://cn.dataone.org/cn/v1/resolve/doi:10.6085%2FAA%2Fmarine_ltm.9.1">
>     <ore:isAggregatedBy rdf:datatype="http://www.w3.org/2001/XMLSchema#anyURI">https://cn.dataone.org/cn/v1/resolve/resourceMap_marine_ltm.9.1#aggregation</ore:isAggregatedBy>
>     <dcterms:identifier rdf:datatype="http://www.w3.org/2001/XMLSchema#string">doi:10.6085/AA/marine_ltm.9.1</dcterms:identifier>
>     <j.0:documents rdf:resource="https://cn.dataone.org/cn/v1/resolve/doi:10.6085%2FAA%2Fmarine_ltm.10.1"/>
>   </rdf:Description>
> ```
> 
> and
> 
> ```
>   <rdf:Description rdf:about="https://cn.dataone.org/cn/v1/resolve/doi:10.6085%2FAA%2Fmarine_ltm.10.1">
>     <ore:isAggregatedBy rdf:datatype="http://www.w3.org/2001/XMLSchema#anyURI">https://cn.dataone.org/cn/v1/resolve/resourceMap_marine_ltm.9.1#aggregation</ore:isAggregatedBy>
>     <dcterms:identifier rdf:datatype="http://www.w3.org/2001/XMLSchema#string">doi:10.6085/AA/marine_ltm.10.1</dcterms:identifier>
>     <j.0:isDocumentedBy rdf:resource="https://cn.dataone.org/cn/v1/resolve/doi:10.6085%2FAA%2Fmarine_ltm.9.1"/>
>   </rdf:Description>
> ```
> 
> This should work with any DataONE metadata format.
> 
> btw - searching for metadata objects with formatId='eml://ecoinformatics' misses EML 2.2.0 objects, as the new formatId for EML 2.2.0 has been updated to 'https://eml.ecoinformatics.org/eml-2.2.0'.
> 
> How would you like to proceed with the PR?

I noticed that those relationships aren't always present in the resource map. For example, in the unit tests I see resource maps that look like [this](https://gist.github.com/ThomasThelen/d6d436073ad3587826c5e47a97c03696). I see that `addMember` adds the relationship _only_ if `addMember` is supplied the science metadata with the data object. This could be a workflow issue on my part, but I only see a few places in the code that supplies `mo`. 

If the SPARQL query fails I think we're out of solid ways for checking what the science metadata is so I'd continue serving the bag withoutit.

ThomasThelen commented:
I reworked the previous commit so that _any_ science metadata is written to the metadata folder. @gothub and @amoeba this should be ready for another review.

amoeba commented:
Re: 

> librdf warning - Variable o was bound but is unused in the query

AFAICT this is not a SPARQL requirement and may just be a spurious thing rdflib does here. I would probably just bind the variable and handle it. Alternatively, if you think it's appropriate to apply here, R has a facility called `suppressWarnings()` which can be useful (sparingly) in situations like this.

ThomasThelen commented:
> Re:
> 
> > librdf warning - Variable o was bound but is unused in the query
> 
> AFAICT this is not a SPARQL requirement and may just be a spurious thing rdflib does here. I would probably just bind the variable and handle it. Alternatively, if you think it's appropriate to apply here, R has a facility called `suppressWarnings()` which can be useful (sparingly) in situations like this.

Fixed in c1f47f1. I originally tried selecting ?o however post-processing became ugly with the XPATH that selects relevant data. In the commit referenced, I just check that ?o is also documented by ?s, which _should_ be true.

amoeba commented on R/DataPackage.R:
```diff
@@ -2226,3 +2211,115 @@ condenseStr <- function(inStr, newLength) {
+    if (isSystemMetadata) {
+        sysmetaXML <- serializeSystemMetadata(objectToWrite, version="v2")
+        objectIdentifier <- objectToWrite@identifier
+        objectPath <- gsub(":", "_", objectPath)
```
This looks not quite safe enough. It looks like `objectPath` can be an identifier, path on the filesystem, or the value provided in `targetPath`. We might sanitize the filesystem path or `targetPath` value elsewhere but, at least, `identifier`s can have characters beyond `:` that need escaping, such as `.`.

amoeba commented on R/DataPackage.R:
```diff
@@ -2226,3 +2211,115 @@ condenseStr <- function(inStr, newLength) {
+        sysmetaXML <- serializeSystemMetadata(objectToWrite, version="v2")
+        objectIdentifier <- objectToWrite@identifier
+        objectPath <- gsub(":", "_", objectPath)
+        fileConn<-file(objectPath)
```
This isn't guaranteed to close. Better to use `writeLines` with a character vector and let it manage the resource than send in a file handle like this. If we really need to be working with file handles, use `on.exit` to ensure the file handle is closed.

amoeba commented on R/DataPackage.R:
```diff
@@ -1588,161 +1571,163 @@ setMethod("serializeToBagIt", signature("DataPackage"), function(x, mapId=NA_cha
+        # Check if the object is a science metadata document by checking if its identifier is in the list of
+        # science metadata identifiers
+        if(in_uri(object_id=object_id, uri_list=metadata_ids)) {
+            science_metadata_filename <- file.path(metadataDir, 'science-metadata.xml')
```
I see we set a file name of `science-metadata.xml` or similar here for the metadata docs in the package. Is this what we decided, as opposed to trusting the user's provided value (i.e., `fileName` in sysmeta)?

ThomasThelen replied:
I don't think we discussed the case where the user sets their own filename.

If we obey the user's filename, we'll have a bug if something tries to read the bag can can't find a `system-metadata.xml` document.
If we don't obey the user's filename, we'll have a bug where the user's file wasn't named as expected.

I think that with a bag we're already committed to splitting the package up and adding files in ways that the user didn't specify (placing their data in a data folder, separating the metadata out into another folder, adding bag files, etc) which makes me want to think that the export is a DataONE thing. The use case for someone downloading a bag is for interchange (whether people actually do this is another question) which makes me want to have a standard name for the science metadata. If we had a format for human consumption (like a plain zip) I'd vote for obeying the user's filename.

amoeba replied:
The BagIt ZIP is totally a thing for human consumption, in fact that's our primary use case for them. I dug up some of our previous conversations on this and am now remembering some of those details so I think this fine at this point.

amoeba commented on R/DataPackage.R:
```diff
@@ -2226,3 +2211,115 @@ condenseStr <- function(inStr, newLength) {
+
+# Writes a data object or system metadata object to a bag
+# When writing system metadata documents, asTag should be true
+write_to_bag <- function(objectToWrite, objectPath, bagDir, isSystemMetadata=FALSE, isScienceMetadata=FALSE) {
```
Rather than setting `isSystemMetadata` and `isScienceMetadata`, can we inspect the type of the object (with `class()`, or `is()`) to know what to do with it? Or, would things be cleaner if `write_to_bag` handled writing the object itself and the System Metadata together, rather than separately?

ThomasThelen replied:
Science metadata objects are the same type as ordinary data objects (DataObject) so we can't use type inference to distinguish the two. System metadata however is definitely distinguishable by type. 

The reason for the ugly complexity is that the science metadata objects share behaviour with data objects and system metadata. They share the same type as plain old data, but are written to the tag manifest file (where sysmeta also goes) rather than the normal manifest where plain old data goes.

I _can_ pass the system metadata in with each object (I would prefer this), replacing `isSystemMetadata` with `systemMetadata`.  Or I can do type deduction to get rid of `isSystemMetadata`, but we'll still need to know if it's science metadata when we're handling a plain old DataObject type.

mbjones replied:
Every DataObject is a combination of the bytes for the object, plus SystemMetadata for the object. Seems to me like we should write them in one operation into their two locations, and not be calling this function separately for SystemMetadata.

The distinction between METADATA and DATA and RESOURCE is based on the DataONE formatId (see https://cn.dataone.org/cn/v2/formats), so the client should not have to determine which type it is and where it belongs with these boolean parameters (which seems error prone to me too in that it leaves decisions up to the client that are really part of the package spec).

ThomasThelen replied:
From what I can tell, the SystemMetadata object doesn't track the format type which led me to initially try to parse the format id, but as Peter pointed out was restricted and could be done more reliably by querying the resource map.

I somehow didn't notice that the system metadata is within DataObject-that will definitely simplify this method and the logic calling it.

ThomasThelen commented on R/DataPackage.R:
```diff
@@ -326,8 +326,8 @@ setMethod("addMember", signature("DataPackage"), function(x, do, mo=NA_character
-    if (!is.na(iObj@relativeFilePath)){
-        insertRelationship(x, getIdentifier(iObj), iObj@relativeFilePath, "http://www.w3.org/ns/prov#atLocation")
+    if (!is.na(iObj@targetPath)){
+        insertRelationship(x, getIdentifier(iObj), iObj@targetPath, "http://www.w3.org/ns/prov#atLocation")
```
Good catch; Fixed with a few other places in b2d520f

ThomasThelen commented on R/DataPackage.R:
```diff
@@ -1588,161 +1571,163 @@ setMethod("serializeToBagIt", signature("DataPackage"), function(x, mapId=NA_cha
+    # Convert the value returned from file.info (bytes) into a more
+    # human readable form.
+    # Size is displayed in bytes
+    if(bagBytes < 1024) {
```
It didn't look like we have a convention defined/generally used in the codebase so I went with the SI standard in commit af29d76

ThomasThelen replied:
It actually looks like rdataone does it differently [here](https://github.com/DataONEorg/rdataone/blob/ac7b6c4b14e71aeefad6ea47a35f8131a8608080/R/D1Client.R#L279). I'll revert that commit to conform to what's being done over there

ThomasThelen replied:
Conforms to the dataone convention in b467131

ThomasThelen commented on R/DataPackage.R:
```diff
@@ -2226,3 +2211,115 @@ condenseStr <- function(inStr, newLength) {
+    # was created. If a DataPackage id was not assigned, then create a unique id.
+    if(is.na(id)) {
+        if(is.na(x@sysmeta@identifier) || is.null(x@sysmeta@identifier)) {
+            id <- sprintf("urn:uuid:%s", UUIDgenerate())
```
Went ahead and changed that instance and a few others in the file with 8579161

ThomasThelen commented on inst/tests/test_DataObject.R:
```diff
@@ -57,13 +57,13 @@ test_that("DataObject constructors work", {
-    relativeFilePath="data/rasters/test.csv"
-    do <- new("DataObject", sm, data, filename="test.csv", relativeFilePath=relativeFilePath)
-    expect_that(do@relativeFilePath, equals(relativeFilePath))
+    targetPath="data/rasters/test.csv"
```
Since this is in the tests, I should have actually been using _both_ slashes (to make sure that if someone gives a windows path in atLocation it should download properly in a unix environment. Changed in 
b83bd67

amoeba replied:
Oh, right. 👍

ThomasThelen commented:
> Should tidy up docs for relativeFilePath

You might be looking at an old commit (`relativeFilePath` was changed to a different name). The slot and param docs are using the same language used in that area of the source ("a character string", etc). I can definitely change it but it might look out of place.

I did update some of the documentation in 2d14c63, mostly overhauling the export section to make the usage of `targetPath` more explicit.

> Indicate relativeFilePath is optional in docs

Fixed in 90add9b

> relativePath is taken in as an arg with no validation. Do we want to sanitize it? i.e., what happens when I provide a values like C:\Foo, /Users/me/bar, or ../../baz?

Fixed in 9152a4c and 8223a499203e82a2c9a68cfd6c8b0dc88597dbd9

I couldn't find a library that handles Path validation so I went with a blacklist approach...

> this is bit unclear and has a typo. Phrases like "unnecessary folders" don't readily make sense to me, though I think I know what you mean. Re-working this along with the documentation for the relativeFilePath arg might make this more clear.

When I refactored the vignette I removed this section and kept the targetPath information contained to the Exporting section. I gave a few examples of what good and bad paths look like which should make it more clear to a reader.

New function names like write_to_bag are snake case but existing functions and methods are camel-cased. I think it'd be best to stay consistent within the package.

Fixed! Must have been writing some Python that day ;)

amoeba commented on R/DataObject.R:
```diff
@@ -624,4 +626,13 @@ setMethod("show", "DataObject",
+# Sanitizes a string that should represent a path to a file
+sanitizePath <- function(targetPath) {
+    # List of things that shouldn't be in a path
+    blacklist <- list( '$', '?', '%', '*', ':', '|', '"', '<', '>', ' ', '..')
```
1. Could we use a neutral term here, such as `filter_list`?
2. A few of those tokens are very reasonable things to see in paths so I worry this would cause confusion. Spaces, for example. And `:` is actually a path separator on macOS.

Did we cover this in the spec?

ThomasThelen replied:
1. Definitely
2. We don't cover this in the spec. I surprisingly couldn't find an R package that does this, so I came up with the ruleset above. It needs to work across macos, Windows, and linux. I removed the colon because of the Windows use case where someone puts a drive letter in the path. I could put something in the spec like 'We only support unix style file paths' and revoke anything that isn't unix compliant.

amoeba replied:
Yeah, there sure are a lot of ways to go with this. Maybe best discussed when/if we write up a spec for this (I think we should). I'd be okay with tweaking this side of the code from this PR in another issue.

gothub replied:
I haven't used it myself, but fs::path_sanitize() removes invalid characters from a filename for various OS types.

ThomasThelen replied:
I ran across `fs::path_sanitize()` in my search but it removes file separator characters. One option could be to split the path at each file separator, `path_sanitize` it, and then re-create the path out of it. It _should_ be a cross platform solution.

ThomasThelen commented on R/DataPackage.R:
```diff
@@ -2226,3 +2211,115 @@ condenseStr <- function(inStr, newLength) {
+    # Transform the results into xml
+    doc <- xmlInternalTreeParse(querytResult, asText=TRUE)
+    # Get the solutions of the query
+    rdfStmtNodes <- getNodeSet(doc, "//rs:solution/rdf:Description/rs:binding/rdf:Description/rs:value")
```
Thanks for the pointer about the namespace prefixes not being constant; I went ahead and made this change in 723c0423daf381923cf1dcf673a68ea3f92a3bfd. The resulting code is, IMO, easier to read than the XML parsing. Thakns!

gothub commented on R/DataPackage.R:
```diff
@@ -1588,161 +1571,162 @@ setMethod("serializeToBagIt", signature("DataPackage"), function(x, mapId=NA_cha
+    metadata_ids <- list()
+    for (uri_element in science_metadata_uris) {
+        metadata_ids<- c(metadata_ids, RCurl::curlUnescape(uri_element))
+    }
```
Suggestion -  use `URLdecode` so that you don't have to pull in the RCurl library:
```
# URLDecode each URI - DataONE resolve URLs typically have the identifier portion URLencoded
    metadataIds <- lapply(science_metadata_uris, URLdecode)
```

gothub replied:
Suggestion -  use `URLdecode` so that you don't have to pull in the RCurl library
```suggestion
  ```
# URLDecode each URI - DataONE resolve URLs typically have the identifier portion URLencoded
    metadataIds <- lapply(science_metadata_uris, URLdecode)
```
```

gothub commented on R/DataPackage.R:
```diff
@@ -1588,161 +1571,162 @@ setMethod("serializeToBagIt", signature("DataPackage"), function(x, mapId=NA_cha
+    for (uri_element in science_metadata_uris) {
+        metadata_ids<- c(metadata_ids, RCurl::curlUnescape(uri_element))
+    }
+
```
```suggestion
    # URLDecode each URI - DataONE resolve URLs typically have the identifier portion URLencoded
    metadataIds <- lapply(science_metadata_uris, URLdecode)
```

gothub commented on R/DataPackage.R:
```diff
@@ -1588,161 +1571,162 @@ setMethod("serializeToBagIt", signature("DataPackage"), function(x, mapId=NA_cha
+    scienceMetadataCount <- 1
+    # URLEncode each URI
+    metadataIds <- list()
+    for (uriElement in scienceMetadataUris) {
+        metadataIds<- c(metadataIds, RCurl::curlUnescape(uriElement))
+    }
```
Use `URLdecode` instead of `RCurl::curlUnescape` so that RCurl library isn't a dependency
```suggestion
# URLDecode each URI - DataONE resolve URLs typically have the identifier portion URLencoded
metadataIds <- lapply(science_metadata_uris, URLdecode)
```

gothub commented on R/DataPackage.R:
```diff
@@ -2226,3 +2210,116 @@ condenseStr <- function(inStr, newLength) {
+
+# Finds the identifiers of any science metadata documents in the package by querying
+# a resource map.
+get_science_metadata <- function(resMap) {
+
+retrieveScienceMetadata <- tryCatch(
+    {
+    # Query that finds all subjects that document another object. If ?o is unused, a warning is raised;
+    # use o by performing a sanity check that the object being documented is also documentedBy the science metadata
+    # object.
+    queryString <- 'PREFIX  cito: <http://purl.org/spar/cito/> 
+        SELECT ?s WHERE {
+            ?s cito:documents ?o .
+            ?o cito:isDocumentedBy ?s .
+        }'
+    query <- new("Query", resMap@world, queryString, base_uri=NULL, query_language="sparql", query_uri=NULL)
+    querytResult <- redland::getResults(query, resMap@model, "csv")
+    
+    # Use the csv format when parsing the results
+    outpath <- tempfile()
+    writeLines(getResults(query, resMap@model, "csv"), outpath)
+    # The first row in the file is the 'subject' string from the query; don't read it
+    scinceMetadataUris <- read.csv(outpath, skip = 1)
+
+    # Make sure that there aren't any duplicate identifiers
+    return (unique(scinceMetadataUris))
+    },
+    error=function(cond) {
+        return (list())
+    })
+    return (retrieveScienceMetadata)
+}
```
This function didn't pass the unit test (for me), so here is a slightly refactored version that does.
```suggestion
get_science_metadata_uris <- function(resMap) {

    query_result <- tryCatch( {
       # Query that finds all subjects that document another object. If ?o is unused, a warning is raised;
       # use o by performing a sanity check that the object being documented is also documentedBy the science metadata
       # object.
        queryString <- 'PREFIX  cito: <http://purl.org/spar/cito/> 
            SELECT ?s WHERE {
                ?s cito:documents ?o .
                ?o cito:isDocumentedBy ?s .
            }'
        query <- new("Query", resMap@world, queryString, base_uri=NULL, query_language="sparql", query_uri=NULL)
        queryResult <- redland::getResults(query, resMap@model, "csv")
        freeQuery(query)
        result <- read.csv(text=queryResult, stringsAsFactors = FALSE)
        # Extract the 'subject' column from the relationship csv and place into a list
        # Make sure that there aren't any duplicate identifiers
        science_metadata_uris <- unique(result[,'s'])
    }, error=function(cond) {
        list()
    })
    
    return(query_result)
}
```

gothub commented on R/DataPackage.R:
```diff
@@ -1588,161 +1571,162 @@ setMethod("serializeToBagIt", signature("DataPackage"), function(x, mapId=NA_cha
+    # Get a resource map so that it can be queried for science metadata
+    resMap <- getResourceMap(x)
+    # Check to see if the user supplied any science metadata
+    scienceMetadataUris <- get_science_metadata(resMap)
```
Use slightly more verbose function name
```suggestion
    science_metadata_uris <- get_science_metadata_uris(resMap)
```
